homeassistant/components/tag/__init__.py

```diff
@@ -90,21 +125,231 @@ async def _update_data(self, item: dict, update_data: dict) -> dict:
         # make last_scanned JSON serializeable
         if LAST_SCANNED in update_data:
             data[LAST_SCANNED] = data[LAST_SCANNED].isoformat()
+        if name := data.get(CONF_NAME):
+            entity_id = self.entity_registry.async_get_entity_id(
+                DOMAIN, DOMAIN, data[TAG_ID]
+            )
+            if entity_id := self.entity_registry.async_get_entity_id(
+                DOMAIN, DOMAIN, data[TAG_ID]
+            ):
+                self.entity_registry.async_update_entity(entity_id, name=name)
+            else:
+                self.entity_registry.async_get_or_create(
+                    DOMAIN,
+                    DOMAIN,
+                    data[TAG_ID],
+                    original_name=data.get(CONF_NAME, DEFAULT_NAME),
+                    suggested_object_id=slugify(data.get(CONF_NAME, DEFAULT_NAME)),
+                )
```

Comment:
Isn't it an error if we get an update for an unknown tag?

Reply:
Actually yes.
The only way this could happen would be if the user somehow is able to remove the tag from the entity registry but it's still in the storage and then scans it.
Should we raise instead instead of recreating the entity?

Reply:
Yeah, just raise